[docs] Add preliminary VT extensions documentation and a little cleanup wrt internals (#1122)
contour-terminal/contour: docs/vt-extensions/synchronized-output.md

## Synchronized Output

Synchronized output is merely implementing the feature as inspired by iTerm2 synchronized output,
except that it's not using the rare DCS but rather the well known `SM ?` and `RM ?`. iTerm2 has now also adopted to use the new syntax instead of using DCS.

### Semantics

When rendering the screen of the terminal,
the Emulator usually iterates through each visible grid cell and renders its current state.

> With applications updating the screen a at higher frequency this can cause tearing.

Reviewer: ```suggestion
With applications updating the screen at a high frequency, this can cause tearing.
```

This mode attempts to mitigate that.

When the synchronization mode is enabled following render calls will keep rendering the last rendered state.
The terminal Emulator keeps processing incoming text and sequences.

> When the synchronized update mode is disabled again the renderer may fetch the latest screen buffer state again,

Reviewer: ```suggestion
When the synchronized update mode is disabled, the renderer may fetch the latest screen buffer state,
```

> effectively avoiding the tearing effect by unintentionally rendering in the middle a of an application screen update.

Reviewer: ```suggestion
effectively avoiding the tearing effect by unintentionally rendering in the middle of an application screen update.
```

See [DECRQM](https://vt100.net/docs/vt510-rm/DECRQM.html) (request)
and [DECRPM](https://vt100.net/docs/vt510-rm/DECRPM.html) (response) for more details.

[DECRPM](https://vt100.net/docs/vt510-rm/DECRPM.html) can respond with different values.

Value | Documentation | Relevance for synchronized output mode ?2026
------|---------------|--------------------------
`0`   | Mode is not recognized | not supported
`1`   | Set   | supported and screen updates are not shown to the user until mode is disabled
`2`   | Reset | supported and screen updates are shown as usual (e.g. as soon as they arrive)
`3`   | Permanently set | undefined
`4`   | Permanently reset | not supported

### Using the feature

Use `CSI ? 2026 h` to enable batching output commands into a command queue.

Use `CSI ? 2026 l` when done with your current frame rendering, implicitly updating the render state by reading out the latest grid buffer state.

### Notation

Some developers name the beginning and end of such a synchronized frame (and therefore the instance)

* `BSU` (begin synchronized update, `CSI ? 2026 h`), and
* `ESU` (end synchronized update, `CSI ? 2026 l`).

### Timeout

> So far there is no real concensus if and if so how long a timeout should be.

Reviewer: ```suggestion
So far there is no real consensus on weather a timeout should be and, if so, for how long.
```
